File: chapter_3/js/src/generic.js
class Generic {
  static get = (data, path) => {
    var res = data;
    if (path) {
      for (var i = 0; res && i < path.length; i++) {
        const key = path[i];
        res = res[key];
      }
      return res;
    } else {
      return undefined;
    }
  };

  static map = (data, mapper) => {
    var result = [];
    if (data) {
      if (data instanceof Array) {
        result = mapper ? data.map(mapper) : data;
      } else if (data instanceof Object) {
        result = Object.keys(data).map((key) =>
          mapper ? undefined : data[key]
        );
      }
    }
    return result;
  };
}

export default Generic;

File: chapter_3/java_with_types/java_with_types/src/main/java/generic/Generic.java
package generic;

import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Stream;

public final class Generic {
    private Generic() {
    }

    public static <T, R> Optional<R> get(T data) {
        return get(data, null);
    }

    public static <T, R> Optional<R> get(T data, Function<InfoPath<T>, InfoPath<R>> extractor) {
        return Optional.ofNullable(extractor).flatMap(e -> e.apply(InfoPath.of(data)).findFirst());
    }

    public interface InfoPath<T> {

        static <T> InfoPath<T> of(T data) {
            return () -> Stream.of(data);
        }

        Stream<T> stream();

        default <R> InfoPath<R> map(Function<? super T, ? extends R> mapper) {
            return () -> this.stream().map(mapper);
        }

        default <K, R> InfoPath<R> map(K key, Class<R> valueType) {
            return () -> this.stream().mapMulti((data, consumer) -> {
                if (data instanceof Map<?, ?> m) {
                    final Object value = m.get(key);
                    if (value != null) {
                        consumer.accept(valueType.cast(value));
                    }
                }
            });
        }

        default Optional<T> findFirst() {
            return this.stream().findFirst();
        }
    }
}
